experimental/dds/tree2/src/core/tree/cursor.ts

```diff
@@ -341,16 +344,19 @@ export function mapCursorFields<T, TCursor extends ITreeCursor = ITreeCursor>(
 
 /**
  * @param cursor - cursor at a node whose fields will be visited.
- * @param f - For on each field.
+ * @param f - A function to run for each field.
+ * If `f` returns `Stop`, the iteration will end.
  * If `f` moves cursor, it must put it back to where it was at the beginning of `f` before returning.
  */
 export function forEachField<TCursor extends ITreeCursor = ITreeCursor>(
 	cursor: TCursor,
-	f: (cursor: TCursor) => void,
+	f: (cursor: TCursor) => void | Stop,
 ): void {
 	assert(cursor.mode === CursorLocationType.Nodes, 0x411 /* should be in nodes */);
 	for (let inField = cursor.firstField(); inField; inField = cursor.nextField()) {
-		f(cursor);
+		if (f(cursor) === Stop) {
+			return;
```

Comment:
You need to `cursor.exitField()` here.

Also, apparently you need to extend the test suite for this function too.

Reply:
Good eye - fixed both cases and added some testing.